src/mixins/canvas_grouping.mixin.js

```diff
@@ -13,8 +13,12 @@
      */
     _shouldGroup: function(e, target) {
       var activeObject = this._activeObject;
-      return activeObject && this._isSelectionKeyPressed(e) && target && target.selectable && this.selection &&
-            (activeObject !== target || activeObject.type === 'activeSelection') && !target.onSelect({ e: e });
+      return !!activeObject && this._isSelectionKeyPressed(e)
+        && !!target && target.selectable && this.selection
+        && (activeObject !== target || activeObject.type === 'activeSelection')
+        //  make sure `activeObject` and `target` aren't ancestors of each other
+        && !target.isDescendantOf(activeObject) && !activeObject.isDescendantOf(target)
+        && !target.onSelect({ e: e });
```

Comment:
ok this boolean condition was unaceppable before and is now too. we need to write this differently or with a lot of comments. I want to remember what is going on here and write it down in a paragraph

Reply:
this is a task for me.